### PR TITLE
PyG Temporal Integration

### LINKED ISSUE #335: PyG Temporal Integration

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #335 

This is a rough sketch of how we can natively support spatio-temporal graphs in TGM.

Notes:
- I managed to bypass redudunt graph copies by creating a bunch of edge events at time=0 containing the static graph. I materialize this edge index once and cache it model side. All remaining events are node only (requires some changes in assumptions to our core model - e.g. empty graph support to accomodate). 
- TGCN and GCLSTM are implemented. Our end-to-end runtime is roughly on par with `pyg_temporal` on A100. I'd need to double check things before making concrete claims
- This also required using higher-dimensional dynamic node features in order to accomodate time horizon lagged predictions (used in `metra_la` and `pems_bay` datasets). Probably could figure out a way to prevent this change, though it's not something that would ultimately go into core library, hence I am holding off

### CODE REVIEW

File: examples/spatiotemporal/gclstm.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+args = parser.parse_args()
+enable_logging(log_file_path=args.log_file_path)
+
+
+class RecurrentGCN(torch.nn.Module):
+    def __init__(self, node_dim: int, embed_dim: int, K: int = 1) -> None:
+        super().__init__()
+        self.recurrent = GCLSTM(in_channels=node_dim, out_channels=embed_dim, K=K)
+        self.linear = nn.Linear(embed_dim, 1)
+        self.cached_edge_index = None
```
Comment (Jacob-Chmura) [inline examples/spatiotemporal/gclstm.py:47]: Cache the static graph edge index from t=0

File: examples/spatiotemporal/gclstm.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        batch: DGBatch,
+        node_feat: torch.tensor,
+        h: torch.Tensor | None = None,
+        c: torch.Tensor | None = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        edge_index = self._get_cached_edge_index(batch)
+        edge_weight = batch.edge_feats
+
+        B, N, _ = node_feat.shape
+        node_feat = node_feat.reshape(B * N, -1)
```
Comment (Jacob-Chmura) [inline examples/spatiotemporal/gclstm.py:60]: This is a hack that allows us to easily accomodate lag-horizon signals (e.g. `metr_la`) and normal 1-shot signal prediction (e.g. `chicken_pox`)

File: examples/spatiotemporal/gclstm.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+def train(
+    loader: DGDataLoader, encoder: nn.Module, opt: torch.optim.Optimizer
+) -> Tuple[float, torch.Tensor, torch.Tensor]:
+    encoder.train()
+    total_loss, h_0, c_0 = 0, None, None
+
+    for batch in tqdm(loader):
+        opt.zero_grad()
+
+        node_feats, y_true = (
```
Comment (Jacob-Chmura) [inline examples/spatiotemporal/gclstm.py:93]: Because our dynamic node features encode `(T, V, D + 1, lag)` where T is signal events, V is all nodes on the graph, D + 1 is node features and target, and lag is pred horizon

File: examples/spatiotemporal/gclstm.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            batch.dynamic_node_feats[..., -1, :],
+        )
+
+        out_seq = []
+        for t in range(node_feats.shape[-1]):
+            out_t, h_0, c_0 = encoder(batch, node_feats[:, :, :, t], h_0, c_0)
+            out_seq.append(out_t)
+        y_pred = torch.cat(out_seq, dim=-1)
+
+        # TODO: Need to normalize with mean/std signal
```
Comment (Jacob-Chmura) [inline examples/spatiotemporal/gclstm.py]: They run prediction and evaluation after global z-score standardization. Not too tricky, just skipped for now as it would not have meaningful impact on latency, and it's optional

File: pyproject.toml [hunk 5]
```diff
@@ -53,6 +54,9 @@ build-backend = "flit_core.buildapi"
 [tool.uv.pip]
 torch-backend = "auto"
 
+[tool.uv.sources]
+torch-geometric-temporal = { git = "https://github.com/benedekrozemberczki/pytorch_geometric_temporal.git" }
```
Comment (Jacob-Chmura) [inline pyproject.toml:59]: Had to install from source due to `torch-scatter` issues in uv.

File: tgm/core/_storage/backends/array_backend.py [hunk 6]
```diff
@@ -200,10 +199,10 @@ def get_static_node_feats_dim(self) -> Optional[int]:
             return None
         return self._data.static_node_feats.shape[1]
 
-    def get_dynamic_node_feats_dim(self) -> Optional[int]:
+    def get_dynamic_node_feats_dim(self) -> Optional[Tuple[int]]:
```
Comment (Jacob-Chmura) [inline tgm/core/_storage/backends/array_backend.py:202]: Consider making this change in core library, and renaming to `get_dynamic_node_feats_shape` instead of dim

File: tgm/data/split.py [hunk 7]
```diff
@@ -123,7 +123,6 @@ def apply(self, data: 'DGData') -> Tuple['DGData', ...]:  # type: ignore
                     pretty_number_format(start),
                     pretty_number_format(end),
                 )
-                continue
```
Comment (Jacob-Chmura) [inline tgm/data/split.py:126]: We enable graphs without edge events. In this case, our val/test splits have no edge events (since the entire graph is just created at time=0)
